cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/StatefulSetOperator.java

```diff
@@ -114,45 +91,40 @@ public StatefulSetOperator(Vertx vertx, KubernetesClient client, long operationT
             return Future.succeededFuture();
         } else {
             Future<Void> result = Future.future();
-            log.info("Rolling update of {}/{}: Rolling pod {}", namespace, name, podName);
-            Future<Void> deleted = Future.future();
             Future<CompositeFuture> deleteFinished = Future.future();
-            Watcher<Pod> watcher = new RollingUpdateWatcher(deleted);
-            Watch watch = podOperations.watch(namespace, podName, watcher);
+            log.info("Rolling update of {}/{}: Rolling pod {}", namespace, name, podName);
+
+            // Determine generation of deleted pod
+            Future<Integer> deleted = getGeneration(namespace, podName);
+
             // Delete the pod
-            log.debug("Rolling update of {}/{}: Waiting for pod {} to be deleted", namespace, name, podName);
-            Future podReconcileFuture = podOperations.reconcile(namespace, podName, null);
-            CompositeFuture.join(podReconcileFuture, deleted).setHandler(deleteResult -> {
-                watch.close();
+            Future<ReconcileResult<Pod>> podReconcileFuture = deleted.compose(l -> {
+                log.debug("Rolling update of {}/{}: Waiting for pod {} to be deleted", namespace, name, podName);
+                // null as desired parameter means pod will be deleted
+                return podOperations.reconcile(namespace, podName, null);
+            }).compose(l -> {
+                Future del = podOperations.waitFor(namespace, name, pollingIntervalMs, timeoutMs, (ignore1, ignore2) -> {
+                    // predicate - changed generation means pod has been updated
+                    int newGen = getPodGeneration(podOperations.get(namespace, podName));
+                    log.debug((deleted.result() != newGen) + " vs " + isPodUpToDate(ss, podName));
+                    return deleted.result() != newGen;
+                });
+                log.debug("Rolling pod {} finished", podName);
+                return del;
+            });
+
+            CompositeFuture.join(podReconcileFuture, Future.succeededFuture()).setHandler(deleteResult -> {
```

Comment:
Why are you doing join with `Future.succeededFuture()`?

Reply:
There were two futures and I replaced one with `Future.succeededFuture()` to test my changes.